core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java

```diff
@@ -0,0 +1,404 @@
+    }
+  }
+
+  @Override
+  public void downgradeLastEdge() {
+    Preconditions.checkState(!endsInInode(),
+        "Cannot downgrade last edge when lock list %s ends in an inode", this);
+    Preconditions.checkState(!mEntries.isEmpty(),
+        "Cannot downgrade last edge when the lock list is empty");
+    Preconditions.checkState(endsInWriteLock(),
+        "Cannot downgrade last edge when lock list %s is not write locked", this);
+
+    if (!endsInMultipleWriteLocks()) {
+      EdgeEntry newEntry = createEdgeEntry(lastEdge(), LockMode.READ);
+      removeLastEntry();
+      addEntry(newEntry);
+    }
+  }
+
+  @Override
+  public void downgradeEdgeToInode(Inode inode, LockMode mode) {
+    Preconditions.checkState(!endsInInode(),
+        "Cannot downgrade from an edge write lock to an inode lock when lock list %s "
+            + "already ends in an inode",
+        this);
+    Preconditions.checkState(!mEntries.isEmpty(),
+        "Cannot downgrade from an edge write lock to an inode lock when the lock list is empty");
+    Preconditions.checkState(endsInWriteLock(),
+        "Cannot downgrade from an edge write lock to an inode lock when lock list %s "
+            + "is not write locked",
+        this);
+
+    if (endsInMultipleWriteLocks()) {
+      lockInode(inode, LockMode.WRITE);
+      return;
+    }
+
+    EdgeEntry edgeEntry = createEdgeEntry(lastEdge(), LockMode.READ);
+    InodeEntry inodeEntry = createInodeEntry(inode, mode);
+    removeLastEntry();
+    addEntry(edgeEntry);
+    addEntry(inodeEntry);
+  }
+
+  /**
+   * Locks an edge and creates and entry without adding the entry to the lock list yet.
+   *
+   * @return the edge entry
+   */
+  private EdgeEntry createEdgeEntry(Edge edge, LockMode mode) {
+    LockResource lock = mInodeLockManager.lockEdge(edge, mode);
+    return new EdgeEntry(lock, edge, mode);
+  }
+
+  /**
+   * Locks an inode and creates and entry without adding the entry to the lock list yet.
+   *
+   * @return the inode entry
+   */
+  private InodeEntry createInodeEntry(Inode inode, LockMode mode) {
+    LockResource lock = mInodeLockManager.lockInode(inode, mode);
+    return new InodeEntry(lock, inode, mode);
+  }
+
+  /**
+   * If the lock list is write locked, returns LockMode.WRITE. Otherwise, returns the given mode.
+   *
+   * This helps us preserve the invariant that there is never a READ lock following a WRITE lock.
+   *
+   * @param mode a lock mode
+   * @return the mode
+   */
+  private LockMode nextLockMode(LockMode mode) {
+    if (endsInWriteLock()) {
+      return LockMode.WRITE;
+    }
+    return mode;
+  }
+
+  /**
+   * Adds an entry to mEntries.
+   *
+   * @param entry the entry to add
+   */
+  private void addEntry(Entry entry) {
+    if (!endsInWriteLock() && entry.getMode() == LockMode.READ) {
+      mFirstWriteLockIndex++;
+    }
+    mEntries.add(entry);
+  }
+
+  /**
+   * Removes and unlocks the last entry from mEntries.
+   */
+  private void removeLastEntry() {
```

Comment:
Are all the calls to operations inside `SimpleInodeLockList` sequential?
Who maintains the synchronization of `SimpleInodeLockList`?